lib/charms/vault_k8s/v0/vault_managers.py

```diff
@@ -504,3 +550,304 @@ def generate_vault_unit_certificate(
 def existing_certificate_is_self_signed(ca_certificate: Certificate) -> bool:
     """Return whether the certificate is a self signed certificate generated by the Vault charm."""
     return ca_certificate.common_name == VAULT_CA_SUBJECT
+
+
+class VaultAutounsealNaming:
+    """Computes the auto-unseal related values for a relation.
+
+    This class is used to compute the static details for a vault-autounseal
+    relation, such as the key name, policy name, and approle name. These values
+    are all based on the relation ID.
+
+    This class provides a central place to manage the naming conventions for
+    the auto-unseal functionality.
+    """
+
+    key_prefix: str = ""
+    policy_prefix: str = "charm-autounseal-"
+    approle_prefix: str = "charm-autounseal-"
+
+    @classmethod
+    def key_name(cls, relation_id) -> str:
+        """Return the key name for the relation."""
+        return f"{cls.key_prefix}{relation_id}"
+
+    @classmethod
+    def policy_name(cls, relation_id) -> str:
+        """Return the policy name for the relation."""
+        return f"{cls.policy_prefix}{relation_id}"
+
+    @classmethod
+    def approle_name(cls, relation_id) -> str:
+        """Return the approle name for the relation."""
+        return f"{cls.approle_prefix}{relation_id}"
+
+
+class VaultAutounsealProviderManager:
+    """Encapsulates the auto-unseal functionality.
+
+    This class provides the business logic for auto-unseal functionality in
+    Vault charms. It is opinionated, and aims to make the interface to enabling
+    and managing the feature as simple as possible. Flexibility is sacrificed
+    for simplicity.
+    """
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        client: VaultClient,
+        provides: VaultAutounsealProvides,
+        ca_cert: str,
+        vault_port: int,
+        mount_path: str = "charm-autounseal",
+    ):
+        self._juju_facade = JujuFacade(charm)
+        self._model = charm.model
+        self._client = client
+        self._provides = provides
+        self._mount_path = mount_path
+        self._ca_cert = ca_cert
+        self._port = vault_port
+
+    def get_address(self, relation: Relation) -> str:
+        """Fetch the address from the relation and return it."""
+        return f"https://{self._juju_facade.get_ingress_address(relation)}:{self._port}"
```

Comment:
It doesn't feel like this getter belongs in this class.

Here we assume that the vault service is available at this ingress address but this behavior (address selection) is not handled by autounseal. Shouldn't we pass in the Vault address as an input parameter (instead of the port)?

Reply:
Sure, we can move this back to the charm, but it pushes a little more logic back there as well, because the address depends on the relation. We can no longer rely on a single `sync()` call.

https://github.com/canonical/vault-k8s-operator/pull/534/commits/b73344ca50ae70673a93f4552c470acdd2fd54f4